### PR TITLE
FT-43 # modernise GeoLocation abstraction

### PR DESCRIPTION
**Pull request by jokeyrhyme**
### Changed

-   **BREAKING**: needs `Promise` support, no longer falls back to `jQuery.Deferred`

-   **BREAKING**: drop AMD / Require.js / UMD support, use CommonJS style and assume a build process

-   **BREAKING**: drop W3C-callback style, return `Promise`s

-   refactor to later support alternative Geolocation drivers (e.g. IP)

### Notes

-   it might be easier to review the commits one-by-one

### CODE REVIEW

File: geolocation.js [hunk 1]
```diff
@@ -60,44 +49,59 @@ function mergePositionOptions (
...
-    onSuccess(clonePosition(position))
-  }, onError, options)
+
+}
+
+function detectDriver () /* : GeolocationDriver | false */ {
+  const availableDrivers = DRIVERS_PREFERENCE.filter((name) => {
+    return DRIVERS[name].isAvailable()
+  })
+  return DRIVERS[availableDrivers[0]] || false
```
Comment (mymattcarroll) [inline geolocation.js]: Could this just be:
```js
return DRIVERS_PREFERENCE.find((name) => DRIVERS[name].isAvailable()) || false
```

File: geolocation.js [hunk 2]
```diff
@@ -60,44 +49,59 @@ function mergePositionOptions (
...
+var DRIVERS_PREFERENCE = [ 'W3C' ]
+
+var DRIVERS /* : { [id:string]: GeolocationDriver } */ = {
+
+  W3C: {
+    isAvailable: function () /* : boolean */ {
+      return !!(
+        typeof navigator !== 'undefined' &&
+        navigator.geolocation &&
+        navigator.geolocation.getCurrentPosition
```
Comment (mymattcarroll) [inline geolocation.js]: do you thinks its worth checking for `typeof === 'function'` here?

**Comment by mymattcarroll** [comment]
LGTM 👍 

Just a couple of comments but nothing that _must_ be changed.

File: geolocation.js (hunk 1 above)
Comment (jokeyrhyme) [reply geolocation.js]: I did a fair bit of fence sitting between ES5 and ES2015 features. I should just commit to ES2015 and be done with it. I'm already using `const` here, so I may as well use `Array#find()`.

File: geolocation.js (hunk 2 above)
Comment (jokeyrhyme) [reply geolocation.js]: Yeah, fair call.

**Pull request merged**